run script in local docker container

b01901143 opened the pull request:
Added README file.

Two running options:
- running directly with commandline.
- running as docker image.

b01901143 committed:
run script in local docker container

spiffxp commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+RUN pip install --no-cache-dir -r requirements.txt
+
+# install kubectl
+RUN curl -LO https://storage.googleapis.com/kubernetes-release/release/`curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt`/bin/linux/amd64/kubectl && \
```
Prefer `$()` to backticks
```suggestion
RUN curl -LO https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl && \
```

spiffxp commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+	mv ./kubectl /usr/local/bin/kubectl
+
+# install gcloud
+RUN echo "deb [signed-by=/usr/share/keyrings/cloud.google.gpg] https://packages.cloud.google.com/apt cloud-sdk main" | tee -a /etc/apt/sources.list.d/google-cloud-sdk.list && \
```
Consider installing from a versioned archive instead of using a distro's package.  This makes it easier to get components.  e.g. https://github.com/kubernetes/test-infra/blob/master/images/gcloud/Dockerfile#L36-L45

Won't block on this

b01901143 replied:
Fixed as suggested.

spiffxp commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+	curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key --keyring /usr/share/keyrings/cloud.google.gpg add - && \
+	apt-get update && apt-get install -y google-cloud-sdk
+
+COPY . .
```
nit: this is going to copy `entrypoint.sh` to `/usr/src/app` when you're also copying it to `/usr/local/bin` below

won't block on this

b01901143 replied:
Avoided copying entrypoint.sh twice.

spiffxp commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+#set authentication
+COPY entrypoint.sh /usr/local/bin/
+RUN ln -s /usr/local/bin/entrypoint.sh / # backwards compat
+RUN export GOOGLE_APPLICATION_CREDENTIALS="/usr/src/app/gcloud_key.json"
```
Prefer the ENV statement for setting env vars (ref: https://docs.docker.com/engine/reference/builder/#env)
```suggestion
ENV GOOGLE_APPLICATION_CREDENTIALS "/usr/src/app/gcloud_key.json"
```

spiffxp commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+RUN ln -s /usr/local/bin/entrypoint.sh / # backwards compat
+RUN export GOOGLE_APPLICATION_CREDENTIALS="/usr/src/app/gcloud_key.json"
+
+ENTRYPOINT ["/entrypoint.sh"]
+
+CMD ["python", "secret-script.py", "--get", "--secret_id=docker-secret"]
```
I'm not sure what the intent is here.  I would suggest using ENTRYPOINT or CMD but not both.

spiffxp replied:
After reading through this PR I would suggest
```suggestion
CMD ["/entrypoint.sh"]
```

spiffxp commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+COPY . .
+
+#set authentication
+COPY entrypoint.sh /usr/local/bin/
+RUN ln -s /usr/local/bin/entrypoint.sh / # backwards compat
```
backwards compat with what? I would probably just 
```suggestion
COPY entrypoint.sh /
```

b01901143 replied:
Agreed. Fixed as suggested.

spiffxp commented on entrypoint.sh:
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
+export GOOGLE_APPLICATION_CREDENTIALS="/usr/src/app/gcloud_key.json"
```
Technically you don't need this because of the ENV in the Dockerfile.  I would follow convention of other env's I've suggested here, and default this if not already set.

b01901143 replied:
I forgot to delete this line.
done.

spiffxp commented on README.md:
I'll need to mount in the service account key somehow I think?

b01901143 replied:
Currently I have the key located locally in the working directory along with other src files.
Thus it would copied to WORKDIR in the container through this line: https://github.com/b01901143/secret-script/blob/fc570896c66c02c7218d0bde14f696dbdfc6c23d/Dockerfile#L28 
Is there a preferred way other than this?

spiffxp replied:
There is; don't store it in the container, mount it in as a volume.  It is generally not a good idea to store secrets inside of a container, because it makes it easy to accidentally leak them, and rotating them requires a rebuild of the container.

```sh
# put service account key in a directory since docker doesn't mount files
mkdir -p service-account && mv spiffxp-sa-key.json service-account

# -v src:dst:mode; in this case
#   - use $(pwd) to generate absolute path for docker
#   - dst doesn't have to already exist
#   - use ro to make volume readonly
# -e ENV=value; in this case demonstrating how I'm overriding the default set in the Dockerfile
docker run -ti --rm \
  -v $(pwd)/service-account:/tmp/service-account:ro \
  -e GOOGLE_APPLICATION_CREDENTIALS=/tmp/service-account/spiffxp-sa-key.json \
  gcr.io/spiffxp-gke-dev/secret-script \
  /bin/bash -c 'gcloud auth activate-service-account --key-file=$GOOGLE_APPLICATION_CREDENTIALS'
# returns 'Activated service account credentials for: [spiffxp-demo-sa@spiffxp-gke-dev.iam.gserviceaccount.com]'
```

spiffxp commented on entrypoint.sh:
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
+export GOOGLE_APPLICATION_CREDENTIALS="/usr/src/app/gcloud_key.json"
+gcloud config set project k8s-jkns-gke-soak
```
I would suggest parameterizing this as an env var with a default, e.g.
```suggestion
GCP_PROJECT="${GCP_PROJECT:-k8s-jkns-gke-soak}"
gcloud config set project "${GCP_PROJECT}"
```

b01901143 replied:
done

spiffxp commented on entrypoint.sh:
```diff
@@ -0,0 +1,19 @@
+export GOOGLE_APPLICATION_CREDENTIALS="/usr/src/app/gcloud_key.json"
+gcloud config set project k8s-jkns-gke-soak
+gcloud config set account shanefu@k8s-jkns-gke-soak.iam.gserviceaccount.com
+gcloud config set compute/zone us-central1-f
```
Same parameterizing on zone here (note: I would put the env var declarations at the top of the file
```suggestion
GCP_ZONE="${GCP_ZONE:-us-central1-f}"
gcloud config set compute/zone ${GCP_ZONE}
```

b01901143 replied:
done

spiffxp commented on entrypoint.sh:
```diff
@@ -0,0 +1,19 @@
+
+export GOOGLE_APPLICATION_CREDENTIALS="/usr/src/app/gcloud_key.json"
+gcloud config set project k8s-jkns-gke-soak
+gcloud config set account shanefu@k8s-jkns-gke-soak.iam.gserviceaccount.com
```
In general I would avoid doing this inside a container.  In this case, you don't need to do this because you're activating the service account below

b01901143 replied:
That is true.
Removed.

spiffxp commented on entrypoint.sh:
```diff
@@ -0,0 +1,19 @@
+gcloud config set project k8s-jkns-gke-soak
+gcloud config set account shanefu@k8s-jkns-gke-soak.iam.gserviceaccount.com
+gcloud config set compute/zone us-central1-f
+gcloud auth activate-service-account --key-file $GOOGLE_APPLICATION_CREDENTIALS
```
Prefer quoting bracing variables (ref: https://google.github.io/styleguide/shellguide.html#variable-expansion)
```suggestion
gcloud auth activate-service-account --key-file "${GOOGLE_APPLICATION_CREDENTIALS}"
```

b01901143 replied:
done

spiffxp commented on entrypoint.sh:
```diff
@@ -0,0 +1,19 @@
+gcloud config set account shanefu@k8s-jkns-gke-soak.iam.gserviceaccount.com
+gcloud config set compute/zone us-central1-f
+gcloud auth activate-service-account --key-file $GOOGLE_APPLICATION_CREDENTIALS
+gcloud container clusters get-credentials shanefu
```
I would parameterize this as well

b01901143 replied:
done

spiffxp commented on entrypoint.sh:
```diff
@@ -0,0 +1,19 @@
+
+kubectl get secrets
+
+python secret-script.py --create --secret_id=docker-secret --file=test-secret.yaml
```
nit: I would make this script executable and invoke it directly
```suggestion
./secret-script.py --create --secret_id=docker-secret --file=test-secret.yaml
```

b01901143 replied:
done

spiffxp commented on secret-script.py:
```diff
@@ -1,14 +1,3 @@
-# Lint as: python3
```
Allow this to be directly executable via whichever `python3` is available first in PATH
```suggestion
#!/usr/bin/env python3
```

b01901143 replied:
Amazed.
Fixed.

spiffxp replied:
Ah right, `./script.py` won't work until you `chmod +x` this file

spiffxp commented on secret-script.py:
```diff
@@ -32,31 +21,61 @@ def main(argv):
 	if len(argv) > 1:
 		raise app.UsageError('Too many command-line arguments.')
 
+	# create secrets with id = FLAGS.secret_id
```
General CLI note: this interface is a little unintuitive.  

It's presented as a bunch of flags but only one of them will take effect, and the order of precedence is somewhat arbitrary.  Consider instead an interface where there is a "verb" that you could use in a case statement, both for execution and further validation.

Also, FWIW, the convention I have seen in kubernetes/test-infra python is to use `argparse` to avoid taking on an external dependency (e.g. https://github.com/kubernetes/test-infra/blob/master/triage/summarize.py#L587-L596)

b01901143 replied:
revised to using an "action" argument including options: "create", "get", "update", and "delete".

spiffxp commented on secret-script.py:
```diff
@@ -32,31 +21,61 @@ def main(argv):
+		gcloudAccessSecretVersion(FLAGS.secret_id, "latest")
+		
+		# sync
+		sync = input("Sync? [y/n] ")
```
Personally I would keep this script non-interactive, but if you find the input useful I would at least ask for a `--force` flag to be able to override it

b01901143 replied:
revised to be non-interactive, always sync'ing

spiffxp reviewed (changes requested):
Some comments and suggestions.  I didn't get as far as trying to build and run this locally since it's hardcoded to point at your project/cluster/acct etc.

spiffxp commented on README.md:
For followup (not this PR) I would recommend naming this after a gcr.io repo you have push access to, since that will be the next step for running on kubernetes, e.g. maybe `gcr.io/k8s-jkns-gke-soak/secret-script` to match the name of the project you're using and the repo the image is built from

b01901143 committed:
Update Dockerfile

use $() instead of backticks

Co-authored-by: Aaron Crickenberger <spiffxp@google.com>

b01901143 committed:
Apply suggestions from code review

Co-authored-by: Aaron Crickenberger <spiffxp@google.com>

b01901143 committed:
modified Dockerfile according to comments

Install cloud-sdk from a versioned archive.
Use CMD to run entrypoint.sh

b01901143 committed:
modify entrypoint.sh

parameterize gcloud config as env variables

b01901143 committed:
modify secret-script.py

revised to using an "action" argument including options: "create", "get", "update", and "delete".
remove interactive input, instead, always syncs secrets on both
platforms.

b01901143 committed:
update image name

b01901143 committed:
add license

spiffxp commented on k8s_utils.py:
```diff
@@ -1,15 +1,4 @@
-
+# k8s secret manager utilities
 from absl import app
 from absl import flags
```
since you're using argparse now, you don't need these absl imports

b01901143 replied:
done

spiffxp commented on gcloud_utils.py:
```diff
@@ -1,15 +1,4 @@
-
+# gcloud secret manager utilities
 from absl import app
 from absl import flags
```
don't need absl imports here as well

b01901143 replied:
removed

spiffxp commented on requirements.txt:
```diff
@@ -0,0 +1,4 @@
+absl-py==0.9.0
```
and you can remove absl from requirements

b01901143 replied:
removed

spiffxp commented on README.md:
What permissions does the service account need? When I ran this with a freshly created service account I ran into a bunch of errors.

b01901143 replied:
Added to README.

Will need container.secrets.* and container.clusterViewer permissions.

spiffxp commented on secret-script.py:
```diff
@@ -1,64 +1,82 @@
+			gcloudAccessSecretVersion(args.secret_id, "latest")
+			
+			# sync
+			print("\n Synchronizing...\n")
```
It would be useful for demo purposes to describe what you're syncing from and to

b01901143 replied:
Add from and to in the synchronizing message.

spiffxp commented on secret-script.py:
```diff
@@ -1,64 +1,82 @@
+	parser.add_argument('--secret_id', help='The id of the secret')
+	parser.add_argument('--file', help='The yaml file containing the secret')
+	parser.add_argument('--direction', help='Options: k2g or g2k')
+	return parser.parse_args(args)
```
FWIW I more commonly see `parser.parse_args()` which eliminates some complexity in the function signature (you could just call `main(parse_args())` instead of passing in a slice, for example)

b01901143 replied:
Didn't notice this. You're right.
Fixed.

spiffxp reviewed (changes requested):
This is pretty close

b01901143 committed:
modified README.md

include instructions for granting permission to service accounts.
remove unnecessary packages

spiffxp commented on README.md:
```suggestion
			- service-secret-role.yaml
```

spiffxp replied:
consider including this file in the repo

b01901143 replied:
Added. Though in this demo I chose the other method (roles/container.developer), because I do not have the permission to add a custom role.

spiffxp commented on README.md:
```suggestion
				    gcloud iam roles create <iam-role-id> --project=k8s-jkns-gke-soak --file=service-secret-role.yaml
```

spiffxp commented on README.md:
```suggestion
				    title: Kubernetes Engine Secret Admin
				    description: Provides access to management of Kubernetes Secrets
```

b01901143 replied:
Much better

spiffxp commented on README.md:
```suggestion
				    gcloud iam roles create <iam-role-id> --project=k8s-jkns-gke-soak --file=service-secret-role.yaml
```

b01901143 replied:
revised

spiffxp commented on README.md:
You'll need `roles/secretmanager.admin` to work with Secret Manager secrets

b01901143 replied:
missed that, added.

spiffxp reviewed (changes requested):
One more tiny round

b01901143 committed:
add custom iam role yaml file

b01901143 committed:
mount service account key

b01901143 committed:
docker run instruction w/ --mount

spiffxp reviewed (approved):
LGTM
